backend/src/rate_limiter/mod.rs

```diff
@@ -0,0 +1,179 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+//! Rate Limiter
+
+use async_trait::async_trait;
+use chrono::TimeDelta;
+use sha2::{Digest, Sha256};
+use sqlx::types::chrono::{DateTime, Utc};
+
+pub(crate) mod provider;
+
+#[derive(Debug, Clone)]
+pub(crate) struct RLConfig {
+    pub(crate) max_requests: u64,
+    pub(crate) time_window: TimeDelta,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct RLKey {
+    key: Vec<u8>,
+}
+
+impl RLKey {
+    pub(crate) fn new(service_name: &str, rpc_name: &str, custom: &[&str]) -> Self {
+        // Generate a key based on the service name
+        let mut inputs = Vec::new();
+        inputs.push(service_name);
+        inputs.push(rpc_name);
+        inputs.extend_from_slice(custom);
+
+        let mut vec = Vec::new();
+        for part in inputs {
```

Comment:
We could avoid creating 2 temporary Vec's, by hashing the length and values immediately.  Also, most likely it makes sense just to store `[u8; 32]` as the `key` field in `RLKey`.